diff --git a/domain.go b/domain.go
index a5d3203..0350243 100644
--- a/domain.go
+++ b/domain.go
@@ -176,17 +176,25 @@ type DomainInterfaceModel struct {
 	Type string `xml:"type,attr"`
 }
 
+type DomainInterfaceSourceAddress struct {
+	Type     string `xml:"type,attr"`
+	Domain   string `xml:"domain,attr"`
+	Bus      string `xml:"bus,attr"`
+	Slot     string `xml:"slot,attr"`
+	Function string `xml:"function,attr"`
+}
+
 type DomainInterfaceSource struct {
-	Bridge  string `xml:"bridge,attr,omitempty"`
-	Dev     string `xml:"dev,attr,omitempty"`
-	Network string `xml:"network,attr,omitempty"`
-	Address string `xml:"address,attr,omitempty"`
-	Type    string `xml:"type,attr,omitempty"`
-	Path    string `xml:"path,attr,omitempty"`
-	Mode    string `xml:"mode,attr,omitempty"`
-	Port    uint   `xml:"port,attr,omitempty"`
-	Service string `xml:"service,attr,omitempty"`
-	Host    string `xml:"host,attr,omitempty"`
+	Bridge  string                        `xml:"bridge,attr,omitempty"`
+	Dev     string                        `xml:"dev,attr,omitempty"`
+	Network string                        `xml:"network,attr,omitempty"`
+	Address *DomainInterfaceSourceAddress `xml:"address"`
+	Type    string                        `xml:"type,attr,omitempty"`
+	Path    string                        `xml:"path,attr,omitempty"`
+	Mode    string                        `xml:"mode,attr,omitempty"`
+	Port    uint                          `xml:"port,attr,omitempty"`
+	Service string                        `xml:"service,attr,omitempty"`
+	Host    string                        `xml:"host,attr,omitempty"`
 }
 
 type DomainInterfaceTarget struct {
@@ -597,6 +605,32 @@ type DomainFeatureList struct {
 	SMM        *DomainFeatureState  `xml:"smm"`
 }
 
+type DomainCPUTuneShares struct {
+	Value uint `xml:",chardata"`
+}
+
+type DomainCPUTunePeriod struct {
+	Value uint64 `xml:",chardata"`
+}
+
+type DomainCPUTuneQuota struct {
+	Value int64 `xml:",chardata"`
+}
+
+type DomainCPUTune struct {
+	Shares *DomainCPUTuneShares `xml:"shares"`
+	Period *DomainCPUTunePeriod `xml:"period"`
+	Quota  *DomainCPUTuneQuota  `xml:"quota"`
+}
+
+type DomainMemoryBackingLocked struct {
+	//empty
+}
+
+type DomainMemoryBacking struct {
+	Locked *DomainMemoryBackingLocked `xml:"locked"`
+}
+
 type DomainQEMUCommandlineArg struct {
 	Value string `xml:"value,attr"`
 }
@@ -616,24 +650,26 @@ type DomainQEMUCommandline struct {
 // matching the order of XML elements that libvirt
 // will generate when dumping XML.
 type Domain struct {
-	XMLName         xml.Name           `xml:"domain"`
-	Type            string             `xml:"type,attr,omitempty"`
-	Name            string             `xml:"name"`
-	UUID            string             `xml:"uuid,omitempty"`
-	Memory          *DomainMemory      `xml:"memory"`
-	CurrentMemory   *DomainMemory      `xml:"currentMemory"`
-	MaximumMemory   *DomainMaxMemory   `xml:"maxMemory"`
-	VCPU            *DomainVCPU        `xml:"vcpu"`
-	Resource        *DomainResource    `xml:"resource"`
-	SysInfo         *DomainSysInfo     `xml:"sysinfo"`
-	OS              *DomainOS          `xml:"os"`
-	Features        *DomainFeatureList `xml:"features"`
-	CPU             *DomainCPU         `xml:"cpu"`
-	Clock           *DomainClock       `xml:"clock,omitempty"`
-	OnPoweroff      string             `xml:"on_poweroff,omitempty"`
-	OnReboot        string             `xml:"on_reboot,omitempty"`
-	OnCrash         string             `xml:"on_crash,omitempty"`
-	Devices         *DomainDeviceList  `xml:"devices"`
+	XMLName         xml.Name             `xml:"domain"`
+	Type            string               `xml:"type,attr,omitempty"`
+	Name            string               `xml:"name"`
+	UUID            string               `xml:"uuid,omitempty"`
+	Memory          *DomainMemory        `xml:"memory"`
+	CurrentMemory   *DomainMemory        `xml:"currentMemory"`
+	MaximumMemory   *DomainMaxMemory     `xml:"maxMemory"`
+	MemoryBacking   *DomainMemoryBacking `xml:"memoryBacking"`
+	VCPU            *DomainVCPU          `xml:"vcpu"`
+	CPUTune         *DomainCPUTune       `xml:"cputune"`
+	Resource        *DomainResource      `xml:"resource"`
+	SysInfo         *DomainSysInfo       `xml:"sysinfo"`
+	OS              *DomainOS            `xml:"os"`
+	Features        *DomainFeatureList   `xml:"features"`
+	CPU             *DomainCPU           `xml:"cpu"`
+	Clock           *DomainClock         `xml:"clock,omitempty"`
+	OnPoweroff      string               `xml:"on_poweroff,omitempty"`
+	OnReboot        string               `xml:"on_reboot,omitempty"`
+	OnCrash         string               `xml:"on_crash,omitempty"`
+	Devices         *DomainDeviceList    `xml:"devices"`
 	QEMUCommandline *DomainQEMUCommandline
 }
 
